token.go

```diff
@@ -0,0 +1,280 @@
+		return nil, fmt.Errorf("unexpected content type: %v", signed.ContentType)
+	}
+	return (*SignedToken)(signed), nil
+}
+
+// Verify verifies the signed token as CMS SignedData.
+// An empty list of KeyUsages in VerifyOptions implies ExtKeyUsageTimeStamping.
+// The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed token.
+// It returns success when the first signer info verification succeeds.
+func (t *SignedToken) Verify(ctx context.Context, opts x509.VerifyOptions) ([]*x509.Certificate, error) {
+	if len(t.SignerInfos) == 0 {
+		return nil, &SignedTokenVerificationError{Msg: "signerInfo not found"}
+	}
+	if len(opts.KeyUsages) == 0 {
+		opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping}
+	}
+	intermediates := x509.NewCertPool()
+	for _, cert := range t.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	signed := (*cms.ParsedSignedData)(t)
+	var lastErr error
+	for _, signerInfo := range t.SignerInfos {
+		signingCertificate, err := t.GetSigningCertificate(&signerInfo)
+		if err != nil {
+			lastErr = &SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		certChain, err := signed.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = &SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		// RFC 3161 2.3: The corresponding certificate MUST contain only one
+		// instance of the extended key usage field extension. And it MUST be
+		// marked as critical.
+		if len(signingCertificate.ExtKeyUsage) == 1 &&
+			signingCertificate.ExtKeyUsage[0] == x509.ExtKeyUsageTimeStamping &&
+			len(signingCertificate.UnknownExtKeyUsage) == 0 {
+			// check if marked as critical
+			for _, ext := range signingCertificate.Extensions {
+				if ext.Id.Equal(oid.ExtKeyUsage) {
+					if ext.Critical {
+						// success verification
+						return certChain, nil
+					}
+					break
+				}
+			}
+			lastErr = &SignedTokenVerificationError{Msg: "signing certificate extended key usage extension must be set as critical"}
+		} else {
+			lastErr = &SignedTokenVerificationError{Msg: "signing certificate must have and only have ExtKeyUsageTimeStamping as extended key usage"}
+		}
+	}
+	return nil, lastErr
+}
+
+// GetSigningCertificate gets the signing certificate identified by SignedToken
+// SignerInfo's SigningCertificate or SigningCertificateV2 attribute.
+// If the IssuerSerial field of signing certificate is missing,
+// use signerInfo's sid instead.
+// The identified signing certificate MUST match the hash in SigningCertificate
+// or SigningCertificateV2.
+// When both are present, SigningCertificateV2 is used because it uses hash
+// algorithm beyond SHA1. If both are missing, an error will be returned.
+//
+// References: RFC 3161 2.4.1 & 2.4.2; RFC 5816; RFC 5035 4; RFC 2634 5.4
+func (t *SignedToken) GetSigningCertificate(signerInfo *cms.SignerInfo) (*x509.Certificate, error) {
```

Comment:
code style: naming: in golang, getters does not have the `Get` prefix.
see also https://go.dev/doc/effective_go#Getters